server/game/core/ability/abilityTargets/CardTargetResolver.js

```diff
@@ -78,15 +78,27 @@ class CardTargetResolver {
             targetResults.delayTargeting = this;
             return;
         }
-        if (context.player.autoSingleTarget) {
-            let legalTargets = this.selector.getAllLegalTargets(context, player);
-            if (legalTargets.length === 1) {
-                context.targets[this.name] = legalTargets[0];
-                if (this.name === 'target') {
-                    context.target = legalTargets[0];
-                }
+
+        const legalTargets = this.selector.getAllLegalTargets(context, player);
+        if (legalTargets.length === 0) {
+            if (context.stage === Stage.PreTarget) {
+                // if there are no targets at the pretarget stage, delay targeting until after costs are paid
```

Comment:
By my reading of the official rules, target selection shouldn't happen until after costs anyways(example in 1.8.9). Could this system be brought in line with that? Would it potentially cause issues where we have to undo paying costs and we don't have the infrastructure for that or something?

Reply:
Discussed in chat